trie.h

```diff
@@ -0,0 +1,37 @@
+#ifndef TRIE_H
+#define TRIE_H
+
+#include <map>
+#include <vector>
+#include <string>
+#include <memory>
+
+using std::string;
+using std::map;
+using std::vector;
+using std::shared_ptr;
+
+class Trie {
+private:
+    struct Node {
+        map<char, shared_ptr<Node>> children;
+        bool value;
+    };
+
+    shared_ptr<Node> head;
+
```

Comment:
You might consider making the Trie class an abstract base class with a creator/factory function or pimpl or something. I'm not super fond of having to look at these 'internal' functions or the internal node class in the public header.